### PR TITLE
Added MySQL HealthChecks Query and Callback support.

### PR DESCRIPTION
**Pull request by turric4n**
**What this PR does / why we need it**:

It improves MySql Health Checks by adding Query and Callback Support.

**Which issue(s) this PR fixes**:

Please reference the issue this PR will close: #_[issue number]_

**Special notes for your reviewer**:

I tried to do my best mantaining original code structure.

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [V ] Code compiles correctly
- [ V] Created/updated tests
- [ V] Unit tests passing
- [ V] End-to-end tests passing
- [ V] Extended the documentation
- [V ] Provided sample for the feature

### CODE REVIEW

**Comment by turric4n** [comment]
@dotnet-policy-service agree [company="{your company}"]

**Comment by turric4n** [comment]
@dotnet-policy-service agree

File: src/HealthChecks.MySql/MySqlHealthCheckOptions.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace HealthChecks.MySql;
+
+/// <summary>
+/// Options for <see cref="MySqlHealthCheck"/>.
+/// </summary>
+public class MySqlHealthCheckOptions
+{
+    /// <summary>
+    /// The MySQL Server connection string to be used.
```
Comment (sungam3r) [inline src/HealthChecks.MySql/MySqlHealthCheckOptions.cs:13]:
```suggestion
    /// The MySQL connection string to be used.
```

**Review by turric4n — COMMENTED** [review]
Right!

File: src/HealthChecks.MySql/MySqlHealthCheck.cs [hunk 2]
```diff
@@ -8,24 +8,38 @@ namespace HealthChecks.MySql;
...
+        _options = options;
     }
 
     /// <inheritdoc />
     public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
     {
         try
         {
-            using var connection = new MySqlConnection(_connectionString);
+            using var connection = new MySqlConnection(_options.ConnectionString);
```
Comment (sungam3r) [inline src/HealthChecks.MySql/MySqlHealthCheck.cs:25]: You forgot to call `MySqlHealthCheckOptions.Configure` here.

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 3]
```diff
@@ -9,13 +10,16 @@ namespace Microsoft.Extensions.DependencyInjection;
 public static class MySqlHealthCheckBuilderExtensions
 {
     private const string NAME = "mysql";
+    internal const string HEALTH_QUERY = "SELECT 1;";
 
     /// <summary>
-    /// Add a health check for MySql databases.
+    /// Add a health check for MySQL/MariaDB Server databases.
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:16]: MariaDB?

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 4]
```diff
@@ -9,13 +10,16 @@ namespace Microsoft.Extensions.DependencyInjection;
...
+    internal const string HEALTH_QUERY = "SELECT 1;";
 
     /// <summary>
-    /// Add a health check for MySql databases.
+    /// Add a health check for MySQL/MariaDB Server databases.
     /// </summary>
     /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
-    /// <param name="connectionString">The MySql connection string to be used.</param>
-    /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'mysql' will be used for the name.</param>
+    /// <param name="connectionString">The Sql Server connection string to be used.</param>
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:19]: copy-paste

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 5]
```diff
@@ -26,14 +30,86 @@ public static class MySqlHealthCheckBuilderExtensions
...
+        return builder.AddMySql(_ => connectionString, healthQuery, configure, name, failureStatus, tags, timeout);
+    }
+
+    /// <summary>
+    /// Add a health check for MySQL/MariaDB Server databases.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+    /// <param name="connectionStringFactory">A factory to build the SQL Server connection string to use.</param>
+    /// <param name="healthQuery">The query to be executed.</param>
+    /// <param name="configure">An optional action to allow additional SQL Server specific configuration.</param>
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:49]: again

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs (hunk 3 above)
Comment (turric4n) [reply src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:16]: yes, there are the same thing (practically)

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 6]
```diff
@@ -26,14 +30,86 @@ public static class MySqlHealthCheckBuilderExtensions
...
+    /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'sqlserver' will be used for the name.</param>
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:50]: Please double-check all xml comments.

Comment (turric4n) [reply src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:50]: ok, you are right

**Comment by sungam3r** [comment]
@turric4n You also need to update API approval file. Just run test locally and replace old file with the new one.

**Comment by cieciurm** [comment]
Hi @turric4n,

Are you planning to finish up this PR this week?
I see there is very small changes required. Perhaps you could make it? 😄

**Comment by turric4n** [comment]
Indeed, I will.

File: src/HealthChecks.MySql/MySqlHealthCheckOptions.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using MySqlConnector;
+
+namespace HealthChecks.MySql;
+
+/// <summary>
+/// Options for <see cref="MySqlHealthCheck"/>.
+/// </summary>
+public class MySqlHealthCheckOptions
```
[the remark this answers is not in the document]
Comment (turric4n) [reply src/HealthChecks.MySql/MySqlHealthCheckOptions.cs:10]: It's ok

**Comment by turric4n** [comment]
@sungam3r @cieciurm I'm passing the tests locally. What I should do now?

**Comment by turric4n** [comment]
Done! I believe I've accounted for everything.

**Comment by sungam3r** [comment]
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1835/files#r1250440967

File: src/HealthChecks.MySql/MySqlHealthCheck.cs (hunk 2 above)
Comment (cieciurm) [reply src/HealthChecks.MySql/MySqlHealthCheck.cs:25]: As mentioned above, I think you're missing something like this:

```csharp
_options?.Configure(connection);
```

Comment (turric4n) [reply src/HealthChecks.MySql/MySqlHealthCheck.cs:25]: Done!

File: src/HealthChecks.MySql/MySqlHealthCheck.cs [hunk 8]
```diff
@@ -8,24 +8,39 @@ namespace HealthChecks.MySql;
...
+            _options.Configure?.Invoke(connection);
             await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
 
+            using var command = connection.CreateCommand();
+
+            command.CommandText = _options.CommandText;
+
+            object? result = await command
+                .ExecuteScalarAsync(cancellationToken)
+                .ConfigureAwait(false);
```
Comment (sungam3r) [inline src/HealthChecks.MySql/MySqlHealthCheck.cs:36]:
```suggestion
            using var command = connection.CreateCommand();
            command.CommandText = _options.CommandText;
            object? result = await command.ExecuteScalarAsync(cancellationToken).ConfigureAwait(false);
```

File: src/HealthChecks.MySql/MySqlHealthCheck.cs [hunk 9]
```diff
@@ -8,24 +8,35 @@ namespace HealthChecks.MySql;
...
+            _options.Configure?.Invoke(connection);
             await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
 
+            using var command = connection.CreateCommand();
+            command.CommandText = _options.CommandText;
+            object? result = await command.ExecuteScalarAsync(cancellationToken).ConfigureAwait(false);
+
+            var returnQueryResults =
+                _options.HealthCheckResultBuilder?
+                    .Invoke(result) ?? HealthCheckResult.Healthy();
```
Comment (sungam3r) [inline src/HealthChecks.MySql/MySqlHealthCheck.cs:36]:
```suggestion
            var returnQueryResults = _options.HealthCheckResultBuilder == null
                ? HealthCheckResult.Healthy()
                : _options.HealthCheckResultBuilder(result);
```

File: src/HealthChecks.MySql/MySqlHealthCheck.cs [hunk 10]
```diff
@@ -8,24 +8,35 @@ namespace HealthChecks.MySql;
...
+            command.CommandText = _options.CommandText;
+            object? result = await command.ExecuteScalarAsync(cancellationToken).ConfigureAwait(false);
+
+            var returnQueryResults = _options.HealthCheckResultBuilder == null
                 ? HealthCheckResult.Healthy()
+                : _options.HealthCheckResultBuilder(result);
+
+            return await connection.PingAsync(cancellationToken).ConfigureAwait(false)
+                ? returnQueryResults
                 : new HealthCheckResult(context.Registration.FailureStatus, description: $"The {nameof(MySqlHealthCheck)} check fail.");
```
Comment (sungam3r) [inline src/HealthChecks.MySql/MySqlHealthCheck.cs:40]:
```suggestion
            return _options.HealthCheckResultBuilder == null
                ? HealthCheckResult.Healthy()
                : _options.HealthCheckResultBuilder(result);
```

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 11]
```diff
@@ -9,12 +10,15 @@ namespace Microsoft.Extensions.DependencyInjection;
...
     private const string NAME = "mysql";
+    internal const string HEALTH_QUERY = "SELECT 1;";
 
     /// <summary>
     /// Add a health check for MySql databases.
     /// </summary>
     /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
     /// <param name="connectionString">The MySql connection string to be used.</param>
+    /// <param name="healthQuery">The query to be executed.</param>
+    /// <param name="configure">An optional action to allow additional MySQL specific configuration</param>
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:21]:
```suggestion
    /// <param name="configure">An optional action to allow additional MySQL specific configuration.</param>
```

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 12]
```diff
@@ -26,14 +30,86 @@ public static class MySqlHealthCheckBuilderExtensions
...
+    }
+
+    /// <summary>
+    /// Add a health check for MySQL databases.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+    /// <param name="connectionStringFactory">A factory to build the MySQL connection string to use.</param>
+    /// <param name="healthQuery">The query to be executed.</param>
+    /// <param name="configure">An optional action to allow additional MySQL specific configuration.</param>
+    /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'sqlserver' will be used for the name.</param>
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:50]:
```suggestion
    /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'mysql' will be used for the name.</param>
```

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 13]
```diff
@@ -9,12 +10,15 @@ namespace Microsoft.Extensions.DependencyInjection;
 public static class MySqlHealthCheckBuilderExtensions
 {
     private const string NAME = "mysql";
+    internal const string HEALTH_QUERY = "SELECT 1;";
 
     /// <summary>
     /// Add a health check for MySql databases.
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:16]:
```suggestion
    /// Add a health check for MySQL databases.
```

File: src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs [hunk 14]
```diff
@@ -9,12 +10,15 @@ namespace Microsoft.Extensions.DependencyInjection;
 public static class MySqlHealthCheckBuilderExtensions
 {
     private const string NAME = "mysql";
+    internal const string HEALTH_QUERY = "SELECT 1;";
 
     /// <summary>
     /// Add a health check for MySql databases.
     /// </summary>
     /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
     /// <param name="connectionString">The MySql connection string to be used.</param>
```
Comment (sungam3r) [inline src/HealthChecks.MySql/DependencyInjection/MySqlHealthCheckBuilderExtensions.cs:19]:
```suggestion
    /// <param name="connectionString">The MySQL connection string to be used.</param>
```

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
Thanks, @turric4n.

**Comment by turric4n** [comment]
Thanks to you all, and sorry for the inconveniences.

**Pull request merged**